SourceCode/src/RayTracer.cpp

```diff
@@ -97,79 +102,180 @@ void RayTracer::render() {
+          Color indirectLightContribution(0, 0, 0);
+          unsigned int sampleSize = DEFAULT_SAMPLE_SIZE;
+          Vector Ny = hitNormal;
+          Vector Nx;
+          Vector Nz;
+          if (std::fabs(Ny[0]) > std::fabs(Ny[1])) {
+            Nx = Vector(Ny[2], 0, -Ny[0]) * (1 / std::sqrtf(Ny[0] * Ny[0] + Ny[2] * Ny[2]));
+          } else {
+            Nx = Vector(0, -Ny[2], Ny[1]) * (1 / sqrtf(Ny[1] * Ny[1] + Ny[2] * Ny[2]));
+          }
+          Nz = Ny * Nx;
+          float probabilityDistributionFunction = 1.0f / (2.0f * PI);
+          for (auto i = 0; i < sampleSize; i++) {
+            float angle1 = RayTracer::distribution(RayTracer::engine);
+            float angle2 = RayTracer::distribution(RayTracer::engine);
+            Vector sample = Vector::getVectorSampleOnHemisphere(angle1, angle2);
+            Vector sampleInWorldCoordinates(  //
+                sample[0] * Nz[0] + sample[1] * Ny[0] + sample[2] * Nx[0],
+                sample[0] * Nz[1] + sample[1] * Ny[1] + sample[2] * Nx[1],
+                sample[0] * Nz[2] + sample[1] * Ny[2] + sample[2] * Nx[2]);
+            indirectLightContribution += angle1 / probabilityDistributionFunction *
+                                         shootRay(Ray(intersectionPoint + sampleInWorldCoordinates * MONTE_CARLO_BIAS,
+                                                      sampleInWorldCoordinates, DiffuseRay),
+                                                  depth + 1);
+            indirectLightContribution = indirectLightContribution * (1.0f / static_cast<float>(sampleSize));
           }
+          finalColor += indirectLightContribution * mesh.material.albedo;
+#endif  // GLOBAL_ILLUMINATION
         }
         return finalColor;
       }
-      case Reflect: {
-        Ray reflection = {intersectionPoint + hitNormal * SHADOW_BIAS, ray.direction.reflect(hitNormal)};
-        finalColor += shootRay(ray, depth + 1) * mesh.material.albedo;
+      case Reflective: {
+        Ray reflectionRay(intersectionPoint + hitNormal * REFLECTION_BIAS,
+                          ray.direction.reflect(hitNormal).getNormalized(), ReflectionRay);
+        Color reflectionColor = shootRay(reflectionRay, depth + 1, mesh.material.ior);
+        finalColor += Color(mesh.material.albedo[0] * reflectionColor[0],  // red
+                            mesh.material.albedo[1] * reflectionColor[1],  // green
+                            mesh.material.albedo[2] * reflectionColor[2]   // blue
+        );
         return finalColor;
       }
+      case Refractive: {
+        // float eta1 = IOR;
+        float eta1 = 1.0f;
+        float eta2 = mesh.material.ior;
+        Vector normal = hitNormal;
+        float incidentDotNormal = ray.direction.dot(normal);
+        if (incidentDotNormal > 0) {
+          std::swap(eta1, eta2);
+          normal = -1 * normal;
+          incidentDotNormal = -incidentDotNormal;
```

Comment:
You do not have to inverse the result from the dot product, only swapping the IORs and inversing the normal is enough. Even that it seems you are handling this inversion properly afterwards I would strongly advise to implement your refraction without it.